### ISSUE TITLE
Testing translation plugins for learn.wordpress.org

### ISSUE THREAD

**Issue by jonathanbossenger** [WordPress/Learn#2231]
As our work progresses on the [Learning Pathways on Learn WordPress](https://make.wordpress.org/training/2023/07/07/project-thread-learning-pathways-on-learn-wordpress/) project, the training team would like to determine a method to manage localized content translation, which is [one of our 2024 team goals](https://make.wordpress.org/training/2024/02/05/training-team-goals-for-2024/).

@Piyopiyo-Kitsune kicked off this effort with a [discussion post on the training team site](https://make.wordpress.org/training/2024/02/07/discussion-translation-plugin-on-learn), asking the team for their suggestions for translation plugin solutions.

Out of that discussion, the following plugins have been identified as possible solutions:

 - [Polylang](https://wordpress.org/plugins/polylang/) - @cynthianorman 
 - [WPML](https://wpml.org/) - @unprintedch

We also identified a backup option:

 - [TranslatePress](https://wordpress.org/plugins/translatepress-multilingual) - @jonathanbossenger 

This issue serves to track the testing of these plugins using the learn.wordpress.org local development environment, both to determine requirements for testing, and to provide feedback on the testing process.

## Individual testing issues (to be created and assigned) 
- [ ] #2284
- [x] #2338
- [x] #2339

**Comment by jonathanbossenger** [comment]
@Piyopiyo-Kitsune @bsanevans @lada7042 I will raise this in the next dev-squad session on Thursday. In the meantime, it would be good if we could gather a list of requirements that we should test for. Any requirements you, or the rest of the team can share would be greatly appreciated.

**Comment by kaitohm** [comment]
Front-end requirements I can think of:

* Language selector **does not show** as country flags. (Many countries speak multiple languages, so describing a language with a flag should be avoided.)
* Languages can be changed without needing to log in to one's WordPress.org account.
* Once a language is selected, it is applied globally.
* Content of the selected language is shown above non-translated content.
* English content is still shown where translations don't exist.

Back-end requirements I can think of:
* Translators do not have permission to create or publish new content.
* Translators have permission to both translate any content, and review content translated by others.
* A flow can be implemented where translations must be approved (reviewed) before published.
* Sensei content types can be translated in the same way as native post types.
* Taxonomies can be translated.
* Translated content is indexed and searchable by Jetpack Search.

**Comment by cynthianorman** [comment]
I use Polylang in my website builds for clients so I can help with testing

**Comment by jonathanbossenger** [comment]
Thank you @cynthianorman I will assign you to the ticket and make a note in the description.

If no one else is available, I will also try and test one of these options.

**Comment by jonathanbossenger** [comment]
Updating to note that I will be testing TranslatePress next week.

**Comment by jennydupuy** [comment]
I can help to test Polylang.

**Comment by cynthianorman** [comment]
Hi @jonathanbossenger or anyone who can assist. Do you have a suggestion on setting up a testing environment for the Polylang plugin? I was expecting the new learn redesign site to be on a testing site with the ability to install the plugin. Also, they have a Pro version. Do we have access to this license or plans to purchase? https://polylang.pro/downloads/polylang-pro/

I think I should add, I'm just now getting myself up to speed with the build of the new Learn site. I assumed it was going to be built using a block theme. Is there an issue on GitHub tracking the build? If we are using the current classic theme then I already have that set up on my local machine so I can test Polylang that way.

**Comment by unprintedch** [comment]
> Back-end requirements I can think of:
> 
> * Translators do not have permission to create or publish new content.
> * Translators have permission to both translate any content, and review content translated by others.

If all this is mandatory, there is no other solution than WPML. Polylang does not have any translator management system.
It is still possible to achieve, but with other solutions like custom code or another user role plugin.

**Comment by cynthianorman** [comment]
> > Back-end requirements I can think of:
> > 
> > * Translators do not have permission to create or publish new content.
> > * Translators have permission to both translate any content, and review content translated by others.
> 
> If all this is mandatory, there is no other solution than WPML. Polylang does not have any translator management system.

Since I'm looking to test Polylang. I just want to be clear here. I automatically think of managing translators with the use of WordPress roles in which may need to be further evaluated.

**Comment by unprintedch** [comment]
I can take care of the testing of wpml. (even if i would recommend polylang.)

Here are some of the wpml capabilites

- [x] Language selector does not show as country flags. (Many countries speak multiple languages, so describing a language with a flag should be avoided.)
- [x] Languages can be changed without needing to log in to one's WordPress.org account.
- [x] Once a language is selected, it is applied globally.
- [ ] Content of the selected language is shown above non-translated content.
_Do you need multiple language on the same page? Or is it only on the grid?_
- [x] English content is still shown where translations don't exist.

- [x] Translators do not have permission to create or publish new content.
- [x] Translators have permission to both translate any content, and review content translated by others.
- [x] A flow can be implemented where translations must be approved (reviewed) before published.
- [ ] Sensei content types can be translated in the same way as native post types.
- [x] Taxonomies can be translated.
- [ ] Translated content is indexed and searchable by Jetpack Search.

I guess the big thing is the sensei compatibility

**Comment by unprintedch** [comment]
Another thought, I heard that multilingual will be in the core. Why don't just keep a custom taxonomy and duplicate posts?
I know it sounds strange, but we could make a filter with this taxonomy and other adjustment, and it could fill the gap until we have an official solution?
And it is the way it seems to be in the testing local env?

**Comment by kaitohm** [comment]
Thanks for testing WPML for the team, @unprintedch 🙌 

> It is still possible to achieve, but with other solutions like custom code or another user role plugin.

https://learn.wordpress.org/ already has a number of custom user roles active on the site. While a solution that doesn't need custom work would be ideal, adding the needed functionality is also an option we can consider.

> Another thought, I heard that multilingual will be in the core. Why don't just keep a custom taxonomy and duplicate posts?

That's exactly what we have right now 😄 The team is experiencing a number of challenges with this method right now, one of those being the complexity of tracking and managing many locales in GitHub and Google docs. As the core multilingual implementation seems to be still a few years away, we're hoping to move to a more robust translation system to fill that period.

**Comment by unprintedch** [comment]
Hey

> https://learn.wordpress.org/ already has a number of custom user roles active on the site. While a solution that doesn't need custom work would be ideal, adding the needed functionality is also an option we can consider.

I'll keep that in mind

> The team is experiencing a number of challenges with this method right now, one of those being the complexity of tracking and managing many locales in GitHub and Google docs.

Could we have examples where are the pain points, Maybe a live chat?

> As the core multilingual implementation seems to be still a **few years away**, we're hoping to move to a more robust translation system to fill that period.

That hurts, it is so important to me! My life could be so much better with a nice solution in core.

In short, the fact is that WPML is rather complete in terms of functionality. 
BUT! It adds much complexity and is sometime capricious, prompt to bugs and leads to frustration. 
I would really take this decision after testing deeply.

About that, could you point me to a very complete lesson that i could use to test?

**Comment by jonathanbossenger** [comment]
@cynthianorman 

```
Do you have a suggestion on setting up a testing environment for the Polylang plugin? I was expecting the new learn redesign site to be on a testing site with the ability to install the plugin.
```

How I planned to do this was to install Polylang as a plugin on our Learn WordPress local development environment, copy one of the current live courses (maybe one of the theme developer courses) from Learn WordPress to the local environment, and test that way.

```
Also, they have a Pro version. Do we have access to this license or plans to purchase? https://polylang.pro/downloads/polylang-pro/
```

Installing Pro versions of plugins might not be possible, but it might be worthwhile mentioning the features that the pro version has as part of your feedback.

```
I think I should add, I'm just now getting myself up to speed with the build of the new Learn site. I assumed it was going to be built using a block theme. Is there an issue on GitHub tracking the build? If we are using the current classic theme then I already have that set up on my local machine so I can test Polylang that way.
```

That process hasn't started yet, and will likely only start once the designs are completed.

**Comment by jonathanbossenger** [comment]
@unprintedch 
```
About that, could you point me to a very complete lesson that i could use to test?
```

At the moment, could I confirm, do you have the local development environment for learn set up?

If you do, then you are able to create courses and lessons locally, to test with those. 

Currently lessons are mostly just an embedded youtube video, and content. Some lessons might have quizzes at the end of them which you can also create locally.

**Comment by unprintedch** [comment]
> @unprintedch
> 
> At the moment, could I confirm, do you have the local development environment for learn set up?

Yes, thanks to you, it was painful but I finally got the all thing working

> If you do, then you are able to create courses and lessons locally, to test with those.
> Currently lessons are mostly just an embedded youtube video, and content. Some lessons might have quizzes at the end of them which you can also create locally.

Ok i guess the quiz are the thing to test?

**Comment by unprintedch** [comment]
@jonathanbossenger when you say:
Installing Pro versions of plugins might not be possible, but it might be worthwhile mentioning the features that the pro version has as part of your feedback.

Is a deal breaker for WPML. There is no free version.

**Comment by jonathanbossenger** [comment]
```
Is a deal breaker for WPML. There is no free version.
```

I will add this to this week's meeting agenda, I'm not sure what the next step would be there.

**Comment by jonathanbossenger** [comment]
@unprintedch could I also ask that you copy your WPML specific feedback to the [Test WPML GitHub issue](https://github.com/WordPress/Learn/issues/2338). Thanks

**Comment by unprintedch** [comment]
I did i little summary there, tell me if you need some specific answers.
Confirm to me if the "nothing is free" is an issue. If not I'll will gladly test it deeply.

**Comment by jonathanbossenger** [comment]
Thank you @unprintedch, much appreciated.

**Comment by kaitohm** [comment]
(I'm moving [Jonathan's comment](https://github.com/WordPress/Learn/issues/2350#issuecomment-2023089080) here so we can keep discussing next steps on this issue.)

The following plugins were tested by the following contributors

- [Polylang](https://wordpress.org/plugins/polylang/) - @cynthianorman
- [WPML](https://wpml.org/) - @unprintedch
- [TranslatePress](https://wordpress.org/plugins/translatepress-multilingual) - @sareiodata

Here is a breakdown of the requirements tested

| Requirement                                                                                         | Polylang | WPML                 | TranslatePress |
|-----------------------------------------------------------------------------------------------------|----------|----------------------|----------------|
| Language selector does not show as country flags.                                                   | x        | x                    | x              |
| Languages can be changed without needing to log in to one's WordPress.org account.                  | x        | x                    | x              |
| Once a language is selected, it is applied globally.                                                | x        | x                    | x              |
| Content of the selected language is shown above non-translated content.                             |          | x                    |                |
| English content is still shown where translations don't exist.                                      | x        | x                    | x              |
|                                                                                                     |          |                      |                |
| Translators do not have permission to create or publish new content.                                | x        | x                    | x              |
| Translators have permission to both translate any content, and review content translated by others. | x        | x                    | x              |
| A flow can be implemented where translations must be approved (reviewed) before published.          | x        | x                    |                |
| Sensei content types can be translated in the same way as native post types.                        | x        |                      | x              |
| Taxonomies can be translated.                                                                       | x        | x                    | x              |
| Translated content is indexed and searchable by Jetpack Search.                                     | x        | needs deeper testing | needs deeper testing               |

General note on Jetpack Search: If it's a external index like the Google search that's found on other parts of wordpress.org, it might work. If it relies on WordPress internally for querying content and sending it to Jetpack servers, it might not. We can only really test this on the live site. 

Additional notes from testers (see linked tickets for testing notes):

[Polylang](https://github.com/WordPress/Learn/issues/2284): 

- It's important to know that we essentially build a separate set of pages/posts/taxonomies and corresponding menu for each language.
- For this reason, we would need to create the copied English page/post and designate it as French (for example) otherwise I don't believe this will work out for us. In other words, the copied English page/post would act as a placeholder until a translator would contribute the translation.

[WPML](https://github.com/WordPress/Learn/issues/2338):

- There is no free version of WPML. 
- In short, the fact is that WPML is rather complete in terms of functionality.
- BUT! It adds much complexity and is sometime capricious, prompt to bugs and leads to frustration.
- There is no (in our experience) other real viable alternative except polylang which is not as complete in terms of features.

[TranslatePress](https://github.com/WordPress/Learn/issues/2339)

- The free version doesn't allow for more than 1 language
- The requirement "Translators do not have permission to create or publish new content" is available in the paid version.
- If you have questions or want a paid version to test out I would be happy to help. (we can create a perpetual license, so you get updates moving forward to the paid version). The paid version is just an addon.

**Comment by kaitohm** [comment]
The team would like to prioritize using GPL compliant plugins. So, let’s focus on the research reports for [Polylang](https://github.com/WordPress/Learn/issues/2284) and [TranslatePress](https://github.com/WordPress/Learn/issues/2339). What are folks' thoughts?

Please leave thoughts below. **Let's wrap up this discussion by April 5th** so we can then work with the Meta Team to get this installed on Learn.

**Comment by digitalchild** [comment]
I've been using TranslatePress on my sites for years. Yes you do need a pro version for multiple language support, but it's extremely lightweight, and the translations move no matter the design. Meaning when we update the site design, as long as the strings are the same, translations do not need to be updated.

**Comment by kaitohm** [comment]
I'm liking the looks of TranslatePress (Pro with add-ons) for two reasons:
* [Translator Accounts](https://translatepress.com/docs/addons/translator-accounts/?utm_source=wp.org&utm_medium=tp-description-page&utm_campaign=TPFree) gives a step to verify first-time contributors. But there is no need to provide back end access for them to translate.
* [DeepL Automatic Translation](https://translatepress.com/docs/addons/deepl-automatic-translation/?utm_source=wp.org&utm_medium=tp-description-page&utm_campaign=TPFree) + the free DeepL API allows us to slowly translate content into the major WordPress languages even if we don't have active translators. (A machine-generated translation is better than no translation.)

**Comment by jonathanbossenger** [comment]
@bsanevans I wonder if now would be a good time to connect with the Meta team and find out what the possibility is of installing TranslatePress and it's Pro add-on? I'd hate to do further testing, only to find out it's not possible to use.

**Comment by kaitohm** [comment]
I have a call with Meta folks about the site redesign next week. I'll bring it up then 📝

**Comment by kaitohm** [comment]
We've started the process of testing TranslatePress in https://github.com/WordPress/Learn/issues/2445

**Comment by sebastienserre** [comment]
Hello,

I'm from the Polylang team and just discover this multilingual project and https://github.com/WordPress/Learn/issues/2284.
As per https://github.com/WordPress/Learn/issues/2231#issuecomment-2024196003, Polylang is checking all your requirements...

Note the DeepL integration is available in Polylang Pro. We also have a "duplicate content" feature which can help to bulk duplicate English content to one or more target languages. This can be helpful to keep the original layout and allow translators, to only translate the content.
Finally `Content of the selected language is shown above non-translated content.` can be done thanks to our content synchronization available in Polylang Pro.

If I'm not too late, let me know how I can help you to better test Polylang. I can also share a Polylang Pro license if Pro feature are needed.

**Comment by sebastienserre** [comment]
Hello,
I've sent this to Slack but, maybe better here.
A coupon code to give a try to Polylang pro for free has just been sent to [learn@wordpress.org](mailto:learn@wordpress.org).
The main difference between Polylang Pro and TranslatePress is Polylang will allow translators to have different translation for a same content. It can help to better match languages particularities
I'm still available on any question.

**Comment by sebastienserre** [comment]
As per @jonathanbossenger advise here is the first feedback on @bsanevans test I've read on this thread:
https://github.com/WordPress/Learn/issues/2231#issuecomment-2024196003
Polylang:

> It's important to know that we essentially build a separate set of pages/posts/taxonomies and corresponding menu for each language.

This is resulting in front end to better performance. Each content has its own ID... this is native for WordPress to query the content.
As a side note, Polylang is filtering `parse_query` to add the language in the main Query. The Language is a term of a custom taxonomy. All is managed in the native WordPress database tables.
It means each improvement of the Performance team on taxonomies, queries... will improve Polylang performance.
 

> For this reason, we would need to create the copied English page/post and designate it as French (for example) otherwise I don't believe this will work out for us. In other words, the copied English page/post would act as a placeholder until a translator would contribute the translation.

By using the Polylang Pro license offered, you will benefit of the Synchronization tool which will replicate the change on a language https://github.com/WordPress/Learn/issues/2231#issuecomment-2024196003
Polylang:
It's important to know that we essentially build a separate set of pages/posts/taxonomies and corresponding menu for each language.
This is resulting in front end to better performance. Each content has it's own ID... this is native for WordPress to query the content. As a side note, Polylang is filtering `parse_query` to add the language in the main Query. The Language is a term of a custom taxonomy. All is managed in the native WordPress database table.
It means each improvement of the Performance team on taxonomies, queries... will improve Polylang performance.
 
For this reason, we would need to create the copied English page/post and designate it as French (for example) otherwise I don't believe this will work out for us. In other words, the copied English page/post would act as a placeholder until a translator would contribute the translation.
By using the Polylang Pro license offered, you will benefit of the Synchronization tool which will replicate the change on a language https://polylang.pro/doc/synchronized-posts/. We also have a [bulk duplicate tool](https://polylang.pro/doc/duplicating-content-in-bulk/) to quickly create a content in a new language.

Out of the box, Polylang (Pro) doesn't have any role or capabilities management. It means if you have the `edit_posts` capabilities you can create content (posts/pages/terms....) in all languages.

[Automatic translation](https://polylang.pro/deepl-machine-translation-now-available-for-polylang-pro/) is recently possible. It came with the latest Polylang Pro version.

**Comment by sebastienserre** [comment]
As per @jonathanbossenger advise, here is the first feedback on @bsanevans test I've read on this thread:
https://github.com/WordPress/Learn/issues/2231#issuecomment-2024196003
Polylang:

> It's important to know that we essentially build a separate set of pages/posts/taxonomies and corresponding menu for each language.

This is resulting in front end to better performance. Each content has its own ID... this is native for WordPress to query the content.
As a side note, Polylang is filtering `parse_query` to add the language in the main Query. The Language is a term of a custom taxonomy. All is managed in the native WordPress database tables.
It means each improvement of the Performance team on taxonomies, queries... will improve Polylang performance.
 
> For this reason, we would need to create the copied English page/post and designate it as French (for example) otherwise I don't believe this will work out for us. In other words, the copied English page/post would act as a placeholder until a translator would contribute the translation.

By using the Polylang Pro license offered, you will benefit of the Synchronization tool which will replicate the change on a language https://polylang.pro/doc/synchronized-posts/. We also have a [bulk duplicate tool](https://polylang.pro/doc/duplicating-content-in-bulk/) to quickly create a content in a new language.

Out of the box, Polylang (Pro) doesn't have any role or capabilities management. It means if you have the `edit_posts` capabilities, you can create content (posts/pages/terms....) in all languages.

[Automatic translation](https://polylang.pro/deepl-machine-translation-now-available-for-polylang-pro/) is recently possible. It came with the latest Polylang Pro version.

**Comment by jonathanbossenger** [comment]
@kathrynwp, update on testing. Currently both plugin options have blockers:

- While the TranslatePress Translator Accounts feature works well, giving us the ability to create specific accounts for translators, the translation functionality works string by string. So it means translators would have to translate sentence by sentence, as opposed to the entire lesson content. Additionally these strings are public as soon as a translator submits them, so there's no approval process. (further reading: https://github.com/WordPress/Learn/issues/2339#issuecomment-2338130694)
- Polylang allows translation of the entire lesson content. However, there is currently a bug with the combination of PolyLang and Sensei that causes linked translated lessons [to no longer have a relationship with their course modules](https://github.com/WordPress/Learn/issues/2284#issuecomment-2359399610). I have asked the representative from PolyLang if they can shed some light on the matter. Additionally, translators would need regular wp-admin access with permissions for at least editing posts, which also means the ability to edit pretty much all content on Learn.

Based on this, we will either have to adapt our process for translation to follow the "string by string" method of TranslatePress or wait to see if PolyLang can figure out the Sensei issue.

**Comment by jonathanbossenger** [comment]
@kathrynwp Sébastien from Polylang as provided a [possible workaround](https://github.com/WordPress/Learn/issues/2284#issuecomment-2360190896) I need to test out. 

I am AFK next week, but I will continue testing when I get back.

**Comment by kathrynwp** [comment]
Thanks so much for your testing so far, @jonathanbossenger 

During some recent discussions with @cynthianorman and @jennydupuy, it seems that having to translate string-by-string would not be an efficient process for the purposes of Learn WP, so that would seem to rule out TranslatePress.

> Out of the box, Polylang (Pro) doesn't have any role or capabilities management. It means if you have the edit_posts capabilities, you can create content (posts/pages/terms....) in all languages.

In terms of limiting access for translators, perhaps we could explore using a role-management plugin alongside Polylang? @sebastienserre Do you have any thoughts around this? Is there a specific role-management plugin you could recommend that you know to be compatible with Polylang?

**Comment by sebastienserre** [comment]
Hello @kathrynwp 
Adding capabilities which could be adding in one or several roles is in our roadmap since a long time... unfortunately, this is not yet in our developers hands.
The only plugin I know is https://wpexpertshub.com/plugins/polylang-user-manager/ but as per their website, it has not been tested for a long time...

**Comment by donnapep** [comment]
In case it's helpful, the Sensei team recently undertook a project to improve compatibility with WPML. There are still a few [known issues](https://github.com/Automattic/sensei/issues?q=is%3Aopen+is%3Aissue+label%3AWPML). Please take a look to see if you would consider any of those to be a dealbreaker.

Besides those there is another issue where, if a student were to switch to a different language midway through a course, their progress would not be synced between those languages. This isn't something we are actively looking to address, as it seems unlikely that the average student will switch between languages while taking a course.

**Comment by jonathanbossenger** [comment]
Thanks everyone. I will be testing the PolyLang fix this week, and I will report back once tested.

**Comment by jonathanbossenger** [comment]
Final PolyLang testing: https://github.com/WordPress/Learn/issues/2284#issuecomment-2459070351

**Comment by jonathanbossenger** [comment]
Update (2024-11-26) we are moving ahead with #2284

**Comment by jonathanbossenger** [comment]
@kaitohm I have created the PR to merge PolyLang Pro. I'm going to close this issue now, and we can continue any further conversation around this in #2284

**Comment by kaitohm** [comment]
Thank you SO MUCH for wrangling this conversation this year, @jonathanbossenger . You're a hero! 🦸‍♂
